index.rst

```diff
@@ -48,14 +48,126 @@
+The batch submission system would check that the user submitting the workflow has permissions to read from the input collections and write to the output collection.
+Workflow executions using the standard infrastructure will never attempt to delete datasets from registry.
+If jobs are being executed with science platform user credentials it will be necessary for the workflow execution system to use the client/server Butler interface (:cite:`DMTN-176`) which may result in an additional bottleneck to the registry and will require significant usage of signed URLs.
 
-.. Add content here.
-.. Do not include the document title (it's automatically added from metadata.yaml).
+.. note::
+  Could rate-limiting registry access be mediated by the Registry REST server to act as a queue?
+
+
+Reducing Registry Access
+========================
+
+When running at scale the datastore reads and writes involving S3 or Google Cloud Storage are robust and load-balancing is handled natively by the cloud infrastructure.
+With a single PostgreSQL registry where all jobs must see an up-to-date view of the registry, reducing connections to the database, the number of transactions running inside the database, and the length of those transactions is critical when attempting to run with hundreds of thousands of simultaneous jobs.
+
+There are many different points at which we can decide to do the registry updates and they are described below.
+
+Immediate
+---------
+
+The current approach used by the ``pipetask`` executor is to call ``butler.get()`` sequentially for each of the requested input datasets (sometimes this read can be deferred, or example when coadding many input images).
+When the ``PipelineTask`` completes each of the output datasets is then stored in the butler with a ``butler.put()``.
+
+It will use the Butler it has been configured to use and in the batch system that is a shared PostgreSQL registry and S3-like object store.
+
+One problem with this approach is that the Butler is designed such that the registry changes and the file writing must all be completed within a database transaction.
+This can result in the database transactions taking a relatively long time for large data files (which first have to be written to local scratch space and then uploaded to the object store).
+Long-lived transactions can cause trouble with database responsiveness when tens of thousands of connections are active.
+
+There are also concerns, as expressed in :jira:`DM-26302`, that the connection management in Butler registry is non-optimal since it disables connection pooling in SQLAlchemy.
+
+.. note::
+  How long does the ``pipetask`` executor keep each registry connection open?
+  Is a single connection used for the duration of execution of each quanta?
+  Or is there one connection used for all the gets at the start and a new connection used for all the puts at the end?
```

Comment:
Right now it's one connection for the duration of at least one quantum.  It's actually one connection for the lifetime of the `Butler` instance, so if we ever wanted to (e.g.) use `InMemoryDatastore`, and hence use longer-lived `Butler`s, those collections would be open even longer.

But my plan for DM-26302 was to instead open a new connection only for each transaction context (that's a wide-reaching refactor, but I think a straightforward one).   Without further changes to `PipelineTask`, that would yield one _SQLAlchemy_ connection for each `get` or `put`, but we'd be using connection pooling, so how many _actual_ connections we'd use would be out of our hands.  We could also make changes to `PipelineTask` that would make it use one transaction for all reads and one transaction for all writes _by default_, but we need to give concrete `PipelineTask`s the ability to override that occasionally, for cases where they can't fit everything in memory.

Reply:
are there cases where a `PipelineTask` author has to do their own `butler.put`? As opposed to using a `DeferredDatasetRef` to delay reading the data (which doesn't need registry involvement).

Reply:
I can at least imagine some "pinch-point" calibration steps that would need to work that way, but I'm not certain any exist (@erykoff, does FGCM need to do multiple `butler.put` calls (for different data IDs) in a single quantum at any point, due to memory pressure?).  I would be nervous about removing that possibility even if we don't have any tasks that exercise it, but if it greatly simplifies some other part of the system we could probably work around such a limitation.

Reply:
Okay. That's fine. With the fully externalized model this is all to a local sqlite registry anyhow. Using connection pooling properly seems like it would be critical for science platform users though (presumably they could create a `Butler` and have it hanging around for hours).

Reply:
I'm not sure what the question is here.  But fgcmcal does do multiple puts in `runQuantum`, one for each visit.  In the latest version (in review) this is done by having the `run` method return a generator that `runQuantum` loops over.  It would be possible to store everything in memory (it's not pretty, but it's less than the amount of memory used in the fit process, so if you're using the same hardware this should be fine), but this seemed to be the right way to do things at the moment.